signer/src/stacks/events.rs

```diff
@@ -215,8 +220,159 @@ fn withdrawal_create(mut map: RawTupleData, network: NetworkKind) -> Result<Regi
+/// human-readable parts is inherited from the network of the underlying
+/// transaction itself (basically, on stacks mainnet we send to mainnet
+/// bitcoin addresses and similarly on stacks testnet we send to bitcoin
+/// testnet addresses).
+///
+/// ## P2PKH
+///
+/// Generally speaking, Pay-to-Public-Key-Hash addresses are formed by
+/// taking the Hash160 of the public key, prefixing it with one byte (0x00
+/// on mainnet and 0x6F on testing) and then base58 encoding the result.
+///
+/// To specify this address type in the `initiate-withdrawal-request`
+/// contract call, the `version` is 0x00 and the `hashbytes` is the Hash160
+/// of the public key.
+///
+///
+/// ## P2SH, P2SH-P2WPKH, and P2SH-P2WSH
+///
+/// Pay-to-script-hash-* addresses are formed by taking the Hash160 of the
+/// locking script, prefixing it with one byte (0x05 on mainnet and 0xC4 on
+/// testnet) and base58 encoding the result. The difference between them
+/// lies with the locking script. For P2SH-P2WPKH addresses, the locking
+/// script is:
+/// ```text
+/// 0 || <Hash160 of the compressed public key>
+/// ```
+/// For P2SH-P2WSH addresses, the locking script is:
+/// ```text
+/// 0 || <sha256 of the redeem script>
+/// ```
+/// And for P2SH addresses you get to choose the locking script in its
+/// entirety.
+///
+/// Again, after you construct the locking script you take its Hash160,
+/// prefix it with one byte and base58 encode it to form the address. To
+/// specify these address types in the `initiate-withdrawal-request`
+/// contract call, the `version` is 0x01, 0x02, and 0x03 (for P2SH,
+/// P2SH-P2WPKH, and P2SH-P2WSH respectively) with the `hashbytes` is the
+/// Hash160 of the locking script.
+///
+///
+/// ## P2WPKH
+///
+/// Pay-to-witness-public-key-hash addresses are formed by creating a
+/// witness program made entirely of the Hash160 of the compressed public
+/// key.
+///
+/// To specify this address type in the `initiate-withdrawal-request`
+/// contract call, the `version` is 0x04 and the `hashbytes` is the Hash160
+/// of the compressed public key.
+///
+///
+/// ## P2WSH
+///
+/// Pay-to-witness-script-hash addresses are formed by taking a witness
+/// program that is compressed entirely of the SHA256 of the redeem script.
+///
+/// To specify this address type in the `initiate-withdrawal-request`
+/// contract call, the `version` is 0x05 and the `hashbytes` is the SHA256
+/// of the redeem script.
+///
+///
+/// ## P2TR
+///
+/// Pay-to-taproot addresses are formed by "tweaking" the x-coordinate of a
+/// public key with a merkle tree. The result of the tweak is used as the
+/// witness program for the address.
+///
+/// To specify this address type in the `initiate-withdrawal-request`
+/// contract call, the `version` is 0x06 and the `hashbytes` is the
+/// "tweaked" public key.
+fn recipient_to_address(mut map: RawTupleData, network: NetworkKind) -> Result<Address, Error> {
```

Comment:
Style suggestion: Given that this is essentially a constructor, I'd consider implementing this using an extension trait on the `Address` type. For example

```
trait FromClarityRecipient {
    from_recipient(map: RawTupleData, network: NetworkKind) -> Result<Self, Error>
}

impl FromClarityRecipient for Address {
    ...
}
```

This way, the code would later read as `Address::from_recipient` which to me reads better than an arbitrary function.

Reply:
Alternatively this could be a method on `RawTupleData` as
```
fn to_address(network: NetworkKind) -> ...
```

Reply:
Good ideas. I think I like the second one more, since I don't see much of a future for the trait approach (what else is going to implement it?). So I'll make it a method for the type.